### PR TITLE
gh-100061: Restore the global Input Stream pointer

### PR DESCRIPTION
**Pull request by uyw4687**
* Issue: gh-100061

### CODE REVIEW

File: Lib/test/test_re.py [hunk 1]
```diff
@@ -2303,6 +2303,8 @@ def test_possessive_quantifiers(self):
         self.assertIsNone(re.match("^x{}+$", "xxx"))
         self.assertTrue(re.match("^x{}+$", "x{}"))
 
+        self.assertEqual(re.match('((.(?!C))++)', 'ABCD').span(), (0, 1))
```
Comment (ghost) [inline Lib/test/test_re.py]: You may add a comment `# gh-100061` for this test-case.

**Review by ghost — COMMENTED** [review]
Please add a new entry that describe the fix to user.
Such as: Fix wrong match in postive qualifiers. Patch by \<your name\>

File: Lib/test/test_re.py (hunk 1 above)
Comment (uyw4687) [reply Lib/test/test_re.py]: Thanks, added :)

File: Modules/_sre/sre_lib.h [hunk 2]
```diff
@@ -1334,6 +1334,11 @@ SRE(match)(SRE_STATE* state, const SRE_CODE* pattern, int toplevel)
                     MARK_POP(ctx->lastmark);
                     LASTMARK_RESTORE();
 
+                    /* Restore the global Input Stream pointer
+                       since it can change after the jump.
+                       See gh-100061 */
+                    state->ptr = ptr;
+
```
Comment (ghost) [inline Modules/_sre/sre_lib.h:1342]: No need mention gh-100061 here.

**Comment by animalize** [comment]
IMO the fix is fine. Just adjust some details (comment/new_entry).
Need to wait for a core developer to merge the PR.

File: Modules/_sre/sre_lib.h (hunk 2 above)
Comment (uyw4687) [reply Modules/_sre/sre_lib.h:1342]: Removed!

**Comment by arhadthedev** [comment]
@ezio-melotti, @serhiy-storchaka (as regex [experts](https://devguide.python.org/core-developers/experts/))

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

Thank you for your contribution @animalize and @uyw4687.

This fixes also #106052

**Pull request merged**